### PR TITLE
[Test][XPU] Refactor `functorch/test_vmap.py` and enable on XPU

### PR DESCRIPTION
**Pull request by pbielak**
- Use XPU function to determine if flash attention is supported
- Refactor `TestVmapAPI`:
  - Add `hw_classification` 
  - Extract device-dependent tests into `TestVmapAPIDevice`
  - Refactor `TestMapAPIDevice` to use `device` parameter and `instantiate_device_type_tests`
  - Enable XPU for `TestVmapAPIDevice`
- Add `hw_classificiation` to `TestVmapOperators`
- Refactor `TestVmapBatchedGradient`
  - add `Device` suffix
  - use `onlyAccelerator`
  - apply XPU specific changes
  - enable on XPU
- Refactor `TestVmapOperatorsOpInfoDevice`
  - apply XPU specific changes
  - enable on XPU
  - update decorators
  - rename class
  - rename entries in `common_method_invocations.py`
- Refactor `TestRandomness`
  - add `Device` suffix
  - add XPU specific changes
  - enable on XPU
- Refactor `TestTransformFailure`
  - add `Device` suffix
  - enable for XPU
- Refactor `TestVmapDeviceType`
  - remove the `Type` suffix
  - enable on XPU
- Refactor `TestVmapNestedTensor`
  - add `Device` suffix
  - enable on XPU
- Refactor the `instantiate_...` calls and group them at the end of the file

### CODE REVIEW

File: test/functorch/test_vmap.py [hunk 1]
```diff
@@ -4752,6 +4759,11 @@ def test_vmap_exhaustive(self, device, dtype, op):
                 # One or more of the overload doesn't have a Batch rule.
                 xfail("bincount"),
                 xfail("torch.ops.aten._scaled_dot_product_flash_attention_for_cpu"),
+                # The vectorized CPU kernels return NaN when the divisor is
+                # FLT_MIN, since Sleef_fmodf8 is only valid for |a / b| <= 2**24.
+                # The scalar CPU path and other backends return the correct value.
+                xfail("fmod", device_type="cpu"),
```
Comment (etaf) [inline test/functorch/test_vmap.py]: These two xfails depend on the CPU vectorization level, so they flip to unexpected successes wherever the vectorized kernel is not used. I checked the divergence on this box (AVX512 machine, float32, divisor = `torch.finfo(torch.float32).tiny`, 64 elements):

```
ATEN_CPU_CAPABILITY=default -> DEFAULT  nan=0   rem_nan=0
ATEN_CPU_CAPABILITY=avx2    -> AVX2     nan=37  rem_nan=37
ATEN_CPU_CAPABILITY=avx512  -> AVX512   nan=37  rem_nan=37
```

`.ci/pytorch/test.sh:397` sets `ATEN_CPU_CAPABILITY=default` for `TEST_CONFIG=nogpu_NO_AVX2`, and `periodic.yml` runs that config over the default test list, which includes `functorch/test_vmap`. There these two tests pass, and an unexpected success is reported as a failure (I confirmed with a throwaway `unittest.expectedFailure` test that passes: pytest under the repo's `pytest.ini` reports `FAILED ... Failed: Unexpected success`). Since periodic does not run on the PR, that breakage would land silently.

Separately, what you found is an eager-mode CPU numerics bug rather than a batch-rule problem: `rhs_make_tensor_kwargs={'exclude_zero': True}` on the `fmod`/`remainder` OpInfos makes `make_tensor` substitute the smallest normal for zero divisors, and the vectorized kernel returns NaN for that divisor while the scalar path and numpy return 0. Please file an issue for the kernel and reference it here, and make the xfail conditional (`active_if` on the CPU capability, or `xfailIf` on the sample) instead of an unconditional `device_type="cpu"` xfail, which drops `fmod`/`remainder` CPU coverage for this test entirely.

File: test/functorch/test_vmap.py [hunk 2]
```diff
@@ -4159,6 +4152,18 @@ def f(q, k, v, dropout):
             fail_with_randomness = randomness == "error"
             if backend != SDPBackend.MATH:
                 fail_with_randomness |= randomness == "same"
+
+            if device_type == "xpu":
```
Comment (etaf) [inline test/functorch/test_vmap.py]: Setting `fail_with_randomness = True` here makes the test assert `assertRaises(RuntimeError)`, but by your own comment the RuntimeError on XPU is "no available kernel", not the randomness error this test exists to check. So the assertion passes for the wrong reason, and would keep passing if vmap's randomness handling regressed on XPU. Use `assertRaisesRegex` with the message you actually expect, or xfail this parametrization with a link to the XPU kernel gap.

For the `EFFICIENT_ATTENTION` case: if XPU's mem-efficient path really just falls back to MATH, the accurate place to say that is `evaluate_platform_supports_efficient_attention()` in `common_cuda.py` (it returns an unconditional `True` for XPU, and you are already editing that file), not a `device_type == "xpu"` branch inside a device-generic test. That would drop this special case instead of adding one.

Also, now that your `common_cuda.py` change requires Xe2+ for flash attention: is the `FLASH_ATTENTION` branch reachable on the XPU CI runner at all? If the runner is PVC, `PLATFORM_SUPPORTS_FLASH_ATTENTION` is now False, flash is not in `PLATFORM_SPECIFIC_SDPA`, and this branch never executes. Which configuration did you see it fail on?

File: torch/testing/_internal/common_cuda.py [hunk 3]
```diff
@@ -195,7 +195,10 @@ def evaluate_platform_supports_flash_attention():
     if TEST_CUDA:
         return not IS_WINDOWS and SM80OrLater
     if TEST_XPU:
-        return True
+        from torch.testing._internal.common_xpu import (
```
Comment (etaf) [inline torch/testing/_internal/common_cuda.py]: `common_xpu` already exposes this predicate as `PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU` (a `LazyVal`), and `test_transformers.py` gates on it. Importing the evaluator function gives two paths to the same thing; prefer the existing constant.

The bigger point is scope. This flips `PLATFORM_SUPPORTS_FLASH_ATTENTION` on XPU from unconditional `True` to `not IS_WINDOWS and Xe2_Or_Later`, which re-gates every test in the repo that reads that flag (`test_transformers.py`, `test_fake_tensor.py`, `test_varlen_attention.py`, several inductor tests), not just vmap. On pre-Xe2 hardware those all start skipping. That is probably the correct behavior, but it is unrelated to the vmap refactor and is invisible in a PR titled "Refactor test_vmap.py" - please split it out, or at minimum spell out in the description which tests it disables and on what hardware.

File: test/functorch/test_vmap.py [hunk 4]
```diff
@@ -1845,7 +1844,7 @@ def get_number(getter):
         test(op, (getter([B0, 2], device), getter([B0], device, torch.double)))
         test(op, (getter([B0], device, torch.double), getter([B0, 2], device)))
 
-        if not torch.cuda.is_available():
+        if not torch.accelerator.is_available():
```
Comment (etaf) [inline test/functorch/test_vmap.py]: Everything this guard protects is commented out - the `return` is followed only by the `TODO(rzou)` block below - so switching `torch.cuda.is_available()` to `torch.accelerator.is_available()` has no effect, while reading as if it gates something. Since you are in here, delete the dead `if`/`return` and leave (or drop) the commented-out TODO rather than updating dead code.

File: test/functorch/test_vmap.py [hunk 5]
```diff
@@ -5131,7 +5143,9 @@ def test_conv_double_backward(self, device):
         is_cuda_sm86 = device.startswith("cuda") and torch.cuda.get_device_capability(
             0
         ) == (8, 6)
-        atol, rtol = (1e-3, 1e-3) if is_cuda_sm86 else (1e-4, 1e-4)
+        atol, rtol = (
+            (1e-3, 1e-3) if is_cuda_sm86 or device.startswith("xpu") else (1e-4, 1e-4)
```
Comment (etaf) [inline test/functorch/test_vmap.py]: `is_cuda_sm86 or device.startswith("xpu")` makes the variable name wrong at the point of use. Name the condition:

```python
needs_loose_tol = is_cuda_sm86 or device.startswith("xpu")
atol, rtol = (1e-3, 1e-3) if needs_loose_tol else (1e-4, 1e-4)
```

And please add a one-line reason for XPU. A 10x tolerance loosening on a conv double-backward numerics test is exactly the kind of thing the next reader will want justified, and unlike `is_cuda_sm86` the XPU case carries no hint of why.

File: test/functorch/test_vmap.py [hunk 6]
```diff
@@ -4070,10 +4064,8 @@ def loss(q, k, v):
         self.assertEqual(actual_backward_grads, expected_backward_grads)
         self.assertEqual(actual_grads, expected_grads)
 
+    @onlyCPU
```
Comment (etaf) [inline test/functorch/test_vmap.py]: This leaves a CPU-only test inside a class labelled `hw_classification = ACCELERATOR`. The filter in `common_utils.get_hw_classification` reads the classification off the class, so `--hw-classification CPU` will not select this test while `ACCELERATOR` will run it on CPU - the inverse of what the `HardwareClassification` docstring says the per-device values are for. Either leave the classification off this class or move the CPU-only sdpa test somewhere it can be labelled `CPU`. The `_cpu` suffix in the method name is also redundant now that the decorator says it.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -12132,8 +12132,8 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
                DecorateInfo(unittest.expectedFailure, 'TestNormalizeOperators', 'test_normalize_operator_exhaustive'),
 
                # vmap: calling random operator not supported
-               DecorateInfo(unittest.skip("Test expects tensor input"), "TestVmapOperatorsOpInfo", "test_vmap_exhaustive"),
-               DecorateInfo(unittest.skip("Test expects tensor input"), "TestVmapOperatorsOpInfo", "test_op_has_batch_rule"),
+               DecorateInfo(unittest.skip("Test expects tensor input"), "TestVmapOperatorsOpInfoDevice", "test_vmap_exhaustive"),
```
Comment (etaf) [inline torch/testing/_internal/common_methods_invocations.py:12284]: The rename missed `torch/testing/_internal/opinfo/definitions/signal.py:332` and `:337`, which still pass `"TestVmapOperatorsOpInfo"`.

File: test/functorch/test_vmap.py (hunk 1 above)
Comment (etaf) [inline test/functorch/test_vmap.py:4765]: Why add this new xfail? the main branch doesn't need these.

**Review by etaf — CHANGES_REQUESTED** [review]

File: test/functorch/test_vmap.py (hunk 1 above)
Comment (pbielak) [reply test/functorch/test_vmap.py:4765]: The class renames triggered this hidden bug - the test class name is used for computing the RNG seed. Before, we luckily avoided sampling zero. Now, a few zero values are sampled - these are prohibited values for these ops, so they get replaced dynamically with `FLT_MIN` (float min). But with these values, the bug described in the comment get triggered, so I added these xfails.

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (pbielak) [reply torch/testing/_internal/common_methods_invocations.py:12284]: Thanks for the catch - must have dropped that by accident during splitting changes into commits.

File: torch/testing/_internal/common_cuda.py (hunk 3 above)
Comment (pbielak) [reply torch/testing/_internal/common_cuda.py]: Thanks - I created a separate PR: https://github.com/pytorch/pytorch/pull/197873

File: test/functorch/test_vmap.py [hunk 8]
```diff
@@ -4159,6 +4152,18 @@ def f(q, k, v, dropout):
...
+
+            if device_type == "xpu":
+                # On XPU, EFFICIENT_ATTENTION currently maps/falls back to the
+                # MATH implementation, so randomness="same" is allowed here.
+                if backend == SDPBackend.EFFICIENT_ATTENTION and randomness == "same":
+                    fail_with_randomness = False
+
+                # On XPU, FLASH_ATTENTION with dropout and randomness="different"
+                # currently has no available kernel, so this path is expected to raise.
+                if backend == SDPBackend.FLASH_ATTENTION and randomness == "different":
```
Comment (etaf) [inline test/functorch/test_vmap.py]: XPU flash attention never accepts dropout. `can_use_flash_attention` in `aten/src/ATen/native/transformers/xpu/sdp_utils.cpp` includes `check_for_dropout`, so the `dropout=0.5` call below fails with "No available kernel. Aborting execution." in every randomness mode, not only "different". I checked this on PVC for error/same/different. That means the XPU flash cases for "error" and "same" also pass because `assertRaises(RuntimeError)` catches the missing-kernel error, not the vmap randomness error. None of the three cases tests randomness handling.

I'd skip the dropout part for XPU flash instead of flipping `fail_with_randomness`:

```python
if device_type == "xpu" and backend == SDPBackend.FLASH_ATTENTION:
    self.skipTest("XPU flash attention does not support dropout")
```

Put it after the `dropout=0.0` call so that part still runs.

File: test/functorch/test_vmap.py [hunk 9]
```diff
@@ -4495,22 +4502,22 @@ def sample_vmap_out_dim_numpy_split_copy_with_int(
         xfail("cdouble"),
         xfail("cfloat"),
         xfail(
-            "jiterator_binary", device_type="cuda"
+            "jiterator_binary", device_type=("cuda", "xpu")
```
Comment (etaf) [inline test/functorch/test_vmap.py]: All the jiterator OpInfos in `common_methods_invocations.py` are decorated with `onlyCUDA`, so on XPU they're skipped before the xfail matters. I ran them on XPU with this branch and every `jiterator_*_xpu` variant reports "Only runs on cuda". Adding `"xpu"` here and to the five jiterator entries in the `test_op_has_batch_rule` xfail set (around L4725-4749) does nothing. Please revert those back to `device_type="cuda"`.

File: test/functorch/test_vmap.py (hunk 5 above)
Comment (etaf) [inline test/functorch/test_vmap.py]: Is the looser tolerance needed on XPU? I reverted this line to the original `1e-4` and `test_conv_double_backward_xpu` still passes on PVC (Max 1100). If BMG needs it, please gate it on that and add a note on why, the same way the sm86 case is gated. Otherwise I'd drop it.

**Review by etaf — APPROVED** [review]

File: test/functorch/test_vmap.py (hunk 8 above)
Comment (pbielak) [reply test/functorch/test_vmap.py]: Thanks for the explanation - I updated the code
